libdd-profiling-ffi/src/profile_status.rs

```diff
@@ -0,0 +1,476 @@
+        } else {
+            Err(status.flags)
+        }
+    }
+}
+
+impl TryFrom<&ProfileStatus> for &CStr {
+    type Error = usize;
+
+    fn try_from(status: &ProfileStatus) -> Result<Self, Self::Error> {
+        if status.flags != FLAG_OK {
+            Ok(unsafe { CStr::from_ptr(status.err.cast_mut()) })
+        } else {
+            Err(status.flags)
+        }
+    }
+}
+
+impl From<ProfileStatus> for Result<(), Cow<'static, CStr>> {
+    fn from(status: ProfileStatus) -> Self {
+        let flags = status.flags;
+        let is_error = (flags & MASK_IS_ERROR) != 0;
+        let is_allocated = (flags & MASK_IS_ALLOCATED) != 0;
+        #[allow(clippy::panic)]
+        if cfg!(debug_assertions) && (status.flags & MASK_UNUSED) != 0 {
+            panic!("invalid bit pattern: {flags:b}");
+        }
+        match (is_allocated, is_error) {
+            (false, false) => Ok(()),
+            (false, true) => Err(Cow::Borrowed(unsafe { CStr::from_ptr(status.err) })),
+            (true, true) => Err(Cow::Owned(unsafe {
+                CString::from_raw(status.err.cast_mut())
+            })),
+            (true, false) => {
+                #[allow(clippy::panic)]
+                if cfg!(debug_assertions) {
+                    panic!("invalid bit pattern: {flags:b}");
+                }
+                unsafe { unreachable_unchecked() }
+            }
+        }
+    }
+}
+
+impl From<()> for ProfileStatus {
+    fn from(_: ()) -> Self {
+        Self::OK
+    }
+}
+
+/// Tries to shrink a vec to exactly fit its length.
+/// On success, the vector's capacity equals its length.
+/// Returns an allocation error if the allocator cannot shrink.
+fn vec_try_shrink_to_fit<T>(vec: &mut Vec<T>) -> Result<(), AllocError> {
+    let len = vec.len();
+    if vec.capacity() == len || core::mem::size_of::<T>() == 0 {
+        return Ok(());
+    }
+
+    // Take ownership temporarily to manipulate raw parts; put an empty vec
+    // in its place.
+    let mut md = ManuallyDrop::new(core::mem::take(vec));
+
+    // Avoid len=0 case for allocators by dropping the allocation and replacing
+    // it with a new empty vec.
+    if len == 0 {
+        // SAFETY: we have exclusive access, and we're not exposing the zombie
+        // bits to safe code since we're just returning (original vec was
+        // replaced by an empty vec).
+        unsafe { ManuallyDrop::drop(&mut md) };
+        return Ok(());
+    }
+
+    let ptr = md.as_mut_ptr();
+    let cap = md.capacity();
+
+    // SAFETY: Vec invariants ensure `cap >= len`, and capacity/len fit isize.
+    let old_layout = unsafe { Layout::array::<T>(cap).unwrap_unchecked() };
+    let new_layout = unsafe { Layout::array::<T>(len).unwrap_unchecked() };
+
+    // SAFETY: `ptr` is non-null and properly aligned for T (Vec invariant).
+    let old_ptr_u8 = unsafe { NonNull::new_unchecked(ptr.cast::<u8>()) };
+
+    match unsafe { Global.shrink(old_ptr_u8, old_layout, new_layout) } {
```

Comment:
It looks like `shrink` is still allowed to panic, even tho that is discouraged https://doc.rust-lang.org/std/alloc/trait.Allocator.html#method.shrink

Reply:
Unfortunately, yes, that's true. Fortunately, we are not using the trait but instead a specific method: https://doc.rust-lang.org/stable/src/alloc/alloc.rs.html#302-307 

It's not really documented, but we can see that it does at least handle _some_ cases and doesn't directly panic.

I mentioned at some point wanting a no-panic "core" for things like this, so we can slap `#[no_panic]` on it and be sure that all paths with panics were optimized away.